Improve scraper german common names 750

temmey opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [x] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
Otherwise please check these points when getting a PR done:
-->

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed the introduction tour
- [ ] I wrote migrations in a way that they are compatible with already present data
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

<!--
Reviewers can copy&check the following to their review.
Also the checklist above can be used.
But also the PR creator should check these points when getting a PR done:
-->

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

temmey commented:
i uploaded a new `mergedDatasets.csv` 8e77a64cfafcedbd76838a4e1e3a90e7188eb4a4f14e14fbec4df37772a8ae55 to nextcloud

markus2330 commented on scraper/src/merge_datasets.js:
```diff
@@ -2,28 +2,95 @@ import fs from "fs";
+                let germanName = title.replace(/ \(.*\)/, "");
+                germanName = germanName.replace('"', "");
+
+                germanName = germanName.replace(unique_name, "");
```
Also unique name with x should be removed, example: `Rosa damascena` with unique name `Rosa x damascena`

Less priority: Passiflora tarminiana (problem seems to be a single occurrence and more difficult to fix)

markus2330 commented on scraper/src/merge_datasets.js:
```diff
@@ -2,28 +2,95 @@ import fs from "fs";
+                  return null;
+                }
+                GermanNamesFound++;
+                plant["common_name_de"] = germanName;
```
For "Allium sativum" it now says Knoblauchöl as German name? Should be Knoblauch.

markus2330 reviewed (changes requested):
Much better but there are still bugs. Would be great to get this fixed for the release today.

markus2330 commented:
Btw. are there some CSV exporter options? It would be great if it can look like Libreoffice export, so that diffing is easier. That means, e.g. no `""` for empty strings. I reformatted mergedDatasets.csv 0ad24c091aebbac2c604847471330303b27cb27d9f1e2bde984924d402575656

temmey commented on scraper/src/merge_datasets.js:
```diff
@@ -2,28 +2,95 @@ import fs from "fs";
+              const entities = data.entities;
+              const entity = entities[id];
+              const dewiki = entity["sitelinks"]["dewiki"];
+
```
"Daucus carota" doesn't have the value "Karotte" in my records, and unfortunately, Wikidata doesn't have it either.
https://www.wikidata.org/w/api.php?action=wbgetentities&sites=enwiki&titles=Daucus%20carota&normalize=&languages=de&format=json

markus2330 replied:
Seems like Wikipedia removed it from their data. So we need a override for this, #726.

markus2330 commented:
Great, thank you, what do we do with the Brandywine Tomato? It is the only case where the unique name is wrong. Maybe simply rename it within the scraper? (It should be `'Brandywine' instead of `(Brandywine)`)

> Btw. are there some CSV exporter options?

Did you find already something?

Did you fix all other points in the review? (https://github.com/ElektraInitiative/PermaplanT/pull/799#discussion_r1285398167 is resolved without answer?)

temmey commented:
> Great, thank you, what do we do with the Brandywine Tomato? It is the only case where the unique name is wrong. Maybe simply rename it within the scraper? (It should be `'Brandywine' instead of `(Brandywine)`)

How do we know that it's wrong? We do have multiple unique names with a (.*). Eg- 

- Solanum melongena 'RS-Me-10.19 (Nala)'
- Capsicum annuum 'RS-Pa-14.20 (Ajvarski)'
- Solanum lycopersicum 'RS-To-02.22 (Indigo Apple)'
- Solanum lycopersicum (Brandywine)


> 
> > Btw. are there some CSV exporter options?
> 
> Did you find already something?

I filter all "" and replace them with null, so no more "" for empty strings.

> 
> Did you fix all other points in the review? ([#799 (comment)](https://github.com/ElektraInitiative/PermaplanT/pull/799#discussion_r1285398167) is resolved without answer?)

 Yes, I fixed the other points.
 ([#799 (comment)](https://github.com/ElektraInitiative/PermaplanT/pull/799#discussion_r1285398167)  I removed since it was unnecessary.

markus2330 commented:
> How do we know that it's wrong? We do have multiple unique names with a (.*). Eg-

The others are okay, the problem with Brandywine is that it doesn't have `'quotes'` which are needed because of our syntax (as described in hierarchy.md).

> I filter all "" and replace them with null, so no more "" for empty strings.

Great. You can test if it round-trips without changes by simply opening in Libreoffice and saving it again.

markus2330 commented:
> Great. You can test if it round-trips without changes by simply opening in Libreoffice and saving it again.

But actually this becomes irrelevant once we have overrides. Then only the overrides file need roundtrip with Libreoffice.

4ydan reviewed (approved):
Looks good except one picky thing maybe, up to you.

markus2330 commented:
- [ ] we don't need the subfamily column
- [ ] Solanum lycopersicum 'Brandywine' not yet fixed
- [ ] common_names_de are now quoted and in square brackets -> should be simply common separated
- [ ] "Zanthoxylum Armatum" and "Zanthoxylum Clavaherculis" are a latin name in common_name_de
- [ ] Cimicifuga racemosa (Actaea racemosa), 'Clover Grass' and  'Kleegras' also have broken names, can be removed

badnames commented on scraper/README.md:
````diff
@@ -89,11 +89,16 @@ This can be done with the following command:
 npm run merge:datasets
 ```
 
+It iteratively gets the German common names from https://www.wikidata.org to prevent 429 errors by sending too many requests in a short time.
````
```suggestion
Iteratively gets the German common names from https://www.wikidata.org to prevent 429 errors by sending too many requests in a short time.
```

badnames commented on scraper/src/merge_datasets.js:
```diff
@@ -2,28 +2,162 @@ import fs from "fs";
+axiosRetry(axios, {
+  retries: 5, // number of retries
+  retryDelay: (retryCount) => {
+    console.log(`retry attempt: ${retryCount}`);
```
It might be worth to consider which HTTP error code was returned.
E.g. we don't need to increase the delay if any other error than 429 comes back.

Similarly, if the target sends status like 404, 401, etc. there is no point in sending another request.
If errors happen often enough this might reduce the scrapers runtime significantly.

(You may ignore this comment if this isn't possible for whatever reason ;) )

markus2330 replied:
@badnames Thx for the reviews, they are appreciated. :heart: 

Maybe this can be implemented at one place for all scrapers? It is quite annoying to get a timeout during scraper runs.

temmey replied:
I implemented it for now to retry only if there is a 429 error, but the other scrapers may need to be reworked so that this would be applied correctly.

badnames commented on scraper/src/merge_datasets.js:
```diff
@@ -291,6 +422,22 @@ async function mergeDatasets() {
 }
 
+/**
+ * Cleans up a JSON array for smoother CSV export.
```
```suggestion
 * Removes empty JSON array entries for smoother CSV export.
```

temmey replied:
This can/should be extended in the future if we want to apply something to all properties of the plants array

markus2330 replied:
I don't really understand your [comment](https://github.com/ElektraInitiative/PermaplanT/pull/799#discussion_r1291966768). Code comments should describe what is, not what might be.

@badnames' comment is much clearer to what was there before (cleaning up an array could be nearly anything). Is @badnames' suggestion not correct? If it is, I would prefer the more clear statement what the code does.

temmey replied:
I extended the function to incorporate the requested changes.

badnames commented on scraper/src/merge_datasets.js:
```diff
@@ -2,28 +2,162 @@ import fs from "fs";
+        .get(
+          `https://www.wikidata.org/w/api.php?action=wbgetentities&sites=enwiki&titles=${unique_name}&normalize=&languages=de&format=json`
+        )
+        .then((response) => {
```
The callback in .then could also be factored out into its own function for improved readability.

badnames reviewed (changes requested):
Good work.
